Issue 10414 - Remove Member Option from View Party link.

titchimoto opened the pull request:
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #10414 

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)

From some initial investigating, it appeared that this issue was occurring due to some permissions issues, because in `index.vue` when it's retrieving the `members-modal` by clicking the 'View Party' link, it's retrieving the `party` from the `this.user` object, which is returning a sort of condensed version of the party/group object? But it doesn't contain some of the details needed for some of the permission checks. 

I've tried to get around this by whenever this link is clicked, it does an `async/await` to retrieve the specific full party details and passes those details to the `members-modal` meaning that those permissions checks for 'Remove Member' etc. can occur and display (i.e only for Group Leaders). This should also fix this issue occurring on other pages like 'Challenges' where the header/index is.

Not being totally familiar with the entire project, I'm not sure if this might cause other issues, but hopefully nothing obvious I'm missing! The `async/await` should only occur when the link is clicked and so shouldn't make too much of an impact at all on performance.

[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID: 0168bfd7-3cc2-4017-92bf-7aeb168e3b8e

titchimoto committed:
Add remove member option from main task page

titchimoto committed:
Code refactor for remove member options

TheHollidayInn commented:
Thanks @titchimoto!

I think we've changed this part recently because of loading the party too many times @paglias, does that sound familiar?

If so, can you add a check to ensure we don't load the party twice?

negue commented:
The change does fix the issue 👍 , but it also loads the party each time you open it from the header. 

Like @TheHollidayInn already mentioned, could you just add a check to see if the party has been already loaded

paglias commented on website/client/components/header/index.vue:
```diff
@@ -175,11 +175,17 @@ export default {
+      // If party, retrieve the full party details to allow permissions checks.
+      if (this.user.party._id) {
+        await this.$store.dispatch('party:getParty', true);
+        this.party = this.$store.state.party.data;
```
is this.party defined only here?

negue replied:
you're right, @titchimoto could you add `party` to the data object?

titchimoto committed:
code refactor to avoid loading party multiple times

titchimoto commented:
@negue @TheHollidayInn Whoops, my bad, I think I accidentally assumed it was already on the data object, added now! :) 

Good point, too, about it loading the party details each time... if I've understood the way it retrieves correctly, I've changed the check to see if there is already a party loaded... if so, totally skip over the `dispatch` and just go straight to opening the modal with the current `party` details. 

Let me know if this works for you if there's anything else obvious I'm missing! 

Cheers :)

paglias commented on website/client/components/header/index.vue:
```diff
@@ -175,11 +176,18 @@ export default {
       }
     },
-    showPartyMembers () {
+    async showPartyMembers () {
```
sorry for the late response @titchimoto as you correctly guessed the problem is because `this.user.party ` that was used here isn't the entire party but just some data for it. 

Just two small notes about the fix:

- `await this.$store.dispatch('party:getParty', true);` by default loads the party again only if it's not been loaded already, while with `true` passed as the second parameter it loads the entire party from the server even if the data is already here.
- So calling `await this.$store.dispatch('party:getParty');` returns the party from memory if it's already been loaded, otherwise it loads it from the server
- There's no need to assign the party to this.party since it's only used in this function

I'd refactor the code removing the `party: {}` inside `data ()` and then the function like this

```
      ...
      // Retrieve the party data
      const party = await this.$store.dispatch('party:getParty');

       // Set the party details for the members-modal component
      this.$store.state.memberModalOptions.groupId = party.data._id;
      ...
      this.$store.state.memberModalOptions.group = party.data;
```

This should work properly without loading the party more than it's needed. I'm not sure if the $dispatch returns the party data directly or inside the `data` property, so please double check that

paglias reviewed (changes requested):
I've left one comment, let me know if it's not clear or tou have any question!

titchimoto committed:
fix dispatch to ensure only pulling once from server

titchimoto commented:
@paglias Awesome! Thanks for all the feedback and suggestions, really helpful & makes much more sense now. Just refactored based on your suggestions and is working fine & much cleaner. Hopefully that should do it! But let me know if there's anything else I can do to improve it. :) Thanks!

negue commented:
@titchimoto could also remove the `Remove Member Option` if its the logged in user? cause nobody can remove itself^^

SabreCat commented:
@negue I think that might be separately fixed on my https://github.com/HabitRPG/habitica/pull/10704 - in which case, maybe this is ready to go?

negue commented:
@SabreCat yes, your PR fixes that issue 👍 

@titchimoto can you resolve the conflicts?

paglias committed:
Merge branch 'develop' into issue-10414-remove-member-from-index

paglias commented:
@titchimoto I'm extremely sorry for the delay here but we've finally managed to merge your PR, thanks!

I've awarded your contributor tier.

Note that further tiers require increasing amounts of work from one to the next, so Tier 2 may require a couple of PRs or a larger PR to attain. But keep helping out and we'll express our gratitude accordingly!
